### DISCUSSION TITLE
Updating text and bargraphs

### DISCUSSION THREAD

**Discussion by yesnoj** [ImpulseAdventure/GUIslice#513] [Q&A]
Hi Paul,
thanks for your piece of software, it's helping on re-write a code.
I've a question about the topic above.
I've a screen where there are some texts and some bargraphs that need to be filled , and text changed after the bargraph is full.
I'm trying to update this components without refreshing all the UI.
For now i'm just testing in the loop(), with a small delay i change text and value in bargraph, but apparently the bargraph/texts blinks every time there is the update.
I can understand that this happens for the text, because is completly different, but why all the bargraph blinks every time i set the new value?

As i wrote above, i'm using this simple sketch for now :

```
void loop()
{
    gslc_Update(&m_gui);

    updateCleaningPage(random(1,7), PUMP_IN_RLY, pumpTimeIn);
    gslc_ElemXProgressSetVal(&m_gui,total_clean_progress,random(100));
    gslc_ElemXProgressSetVal(&m_gui,task_clean_progress,random(100));
    gslc_ElemXProgressSetVal(&m_gui,step_clean_progress,random(100));
    delay(500);
}
```

There is a way to force the update of a single component anytime i want instead of use the method     gslc_Update(&m_gui) that will refresh all the UI?
Thanks in advance!

PS: after some test, i've noticed that the blinking of the bargraph is linked to the string change, in the case the string doesn't change, the bargraph is filled without blink all the border also...mmmm

**Comment by Pconti31** [comment]
@yesnoj If you zip up a small sample project folder and post it here that shows the problem I'll take a look at it.
Paul--

**Reply by yesnoj** [reply]
Ciao Paul, here you have it https://file.io/K8ok5iHuKKFY
I've noticed, if i comment alll the gslc_ElemSetTxtStr works ok...just one gslc_ElemSetTxtStr and do blinks...

**Comment by yesnoj** [comment]
https://github.com/ImpulseAdventure/GUIslice/assets/6757126/df233ab3-aa1d-4f21-b2ce-dfd95b613ae7

This is without ElemSetTxtStr...

**Comment by yesnoj** [comment]
https://github.com/ImpulseAdventure/GUIslice/assets/6757126/0c9aa72b-291c-4537-821b-d2359bae417c

This is updating strings at same time with bargraphs...

**Comment by Pconti31** [comment]
@yesnoj I know the problem   I fixed this in the past by creating my own XProgressbar but it requires using TFT_eSPI instead of Adafruit' GFX drivers.  The reason is that I make use of sprites (something that is used when creating video games).
See my repository for a picture and the code.
[progress-percent](https://github.com/Pconti31/GUIslice-Solutions/tree/main/apps/progress-percent)
Now if you can't use TFT_eSPI let me know and I'll look into a special version for Adafruit's GFX using their canvas interface which is similar to sprites.   It will take day or two and I can't promise it will work.  No matter what I'll post what success or failure occurs.
If failure you will need to wait on Calvin @ImpulseAdventure to think of something.
Anyway, let me know,
Paul--

**Reply by yesnoj** [reply]
Paul good morning.
Use TFT_eSPI instead of Adafruit means what?That i need to change all methods to create buttons ecc ecc? Or just the config file (ard-shld-mcufriend_4wire.h)? 
All my previous code was wrote using the "tft" method to create all the stuff from scratch (tft.fillRect....tft.setTextSize....tft.setCursor....tft.println...).
However i've the ili9341 Mcufriend_kvb, so probably will never work :/
Thanks for your time Paul, I've rewrote over 4k lines of code, if i need to stop, i'll remain with my previous code :(
But if you have spare time, we can do an attempt!
Good sunday,
F.

PS: my idea at first was to use this 2.4" touch display, but in the future to use a bigger one, maybe a 5"...if you have some suggestion...maybe i can move forward (hopefully i don't have to rewrite all the stuff again...) 
In my project i need about 25 digital I/O...for this reason now i'm using the Arduino 2 (however i've accidentally found an ESP32 WROOM 32 devkit with wifi/BLE at home)

**Comment by Pconti31** [comment]
@yesnoj **_Use TFT_eSPI instead of Adafruit means what?_**
Change of Config files.  I believe Bodmer supports your screen although I'm not sure about the Arduino 2 check his repository at
[Bodmer/TFT_eSPI](https://github.com/Bodmer/TFT_eSPI). Although I seem to remember that while he supported mcufriend screens it required soldering or modifying the display which I would suggest you avoid.
 
A esp32 would work nicely with a larger screen since you would have more memory to work with something to think about for the future.

I do hate to see you go through a large re-work of your code though so I'll take a closer look at your sample and see if I can come up with a simple solution with a new XProgress bar to avoid the flicker or blinks.  I do have an idea  for a fix to XProgrees bar, I should know if it works or not with your sample application by the end of today.

If you do switch over to TFT_eSPI post that information and I'll stop working on a new XProgress bar.
Paul--

**Reply by yesnoj** [reply]
Ciao Paul,
"Change of Config files" you mean that i've to chose another file in the GUIslice\configs\? One of the tftespi in the list?! This config file is linked to the display I use, what i need to chose to do a test? 
I need also to find how link the pins of the display to the ESP32-WROOM 32 :/
However if you have a suggestion for ESP display that has pins for connect my relays to it...please suggest me...i can consider to move towards ESP :/

PS: i've found this board, https://www.amazon.it/MYAMIA-Wemos-Esp32-Bluetooth-Bordo/dp/B07G2WDF19 , this maybe can be used to connect my actual display and use the ESP32 available on board...like i do with my Arduino 2.

**Comment by Pconti31** [comment]
@yesnoj I'm sorry but my experiments have failed to improve the display.  In fact even using my TFT_eSPI didn't seem to help much.
I'll have to leave this to Calvin @ImpulseAdventure  to solve.

I'm not sure its worth it to change to TFT_eSPI but if you do so first get Bodmer's samples to work with your display and cpu do not try GUIslice first.   To use TFT_eSPI drivers you must first configure his config file User_Setup.h which is a topic you need to take up over in his repository.

Paul--

**Comment by Pconti31** [comment]
@yesnoj Actually further testing with my XProgress control and TFT_eSPI driver  I discovered why my control performed as badly as the original version.  I had simply tried to run your sample unchanged except for using my control. 

 When I modified your sample to use my new call gslc_ElemXProgressSetFont() to add your text to the control and removed the three text fields (since the text is now inside the control), and also increased the bar size from 10 to 20 to hold the text it all worked perfectly.  No flicker or blinking occurred.

attached is the modified version of your sample. Of course, you can't run it without converting to TFT_eSPI.  
[TestBars.zip](https://github.com/ImpulseAdventure/GUIslice/files/11585399/TestBars.zip)


I can't modify my control to use Adafruit GFX driver because their canvas interface would take up too much memory for your Arduino 2.

Paull--

**Reply by yesnoj** [reply]
Thanks Paul for your time!
To convert to TFT_eSPI i need to use the ESP32 and connect in someway the display that i'm using right now...correct?
Do you have a suggestion for a new hardware to program and solve definetly the issue? :(

**Comment by Pconti31** [comment]
@yesnoj Well, I use products from Adafruit display ILI9341 using SPI but there are lots of good choices.  Here google is your friend use it to do some research.
Here is a good tutorial on Using [tft_espi-library-on-esp32](https://iotips.tech/using-tft_espi-library-on-esp32/)
I'm sure you can find others,
Paul-

**Reply by yesnoj** [reply]
Paul, can i still use GUIslice builder if i do this switch?Or depends from the display i will chose?I've done all this to use your builder...with the idea to manage graphics easily via PC...instead of write coordinates...
F.

**Comment by Pconti31** [comment]
@yesnoj Of course.  If you open the TestBar project I posted above you will notice the only real difference to the Builder is on the E_PROJECT_OPTIONS Tab where "Graphics Library" property has been changed from "Adafruit_GFX" to "TFT_eSPI".
The Builder only cares about what driver you are using and screen size.

Also, you should check to see if TFT_eSPI already supports your screen.  He supports a lot of different displays and interfaces.  If your Arduino 2 is 32 bit TFT_eSPI might also support it and if not maybe Bodmer will support it if you ask.

Paul--

**Reply by yesnoj** [reply]
ciao Paul, i've purchased today the ESP32 https://de.aliexpress.com/item/1005001566084830.html?spm=a2g0o.order_list.order_list_main.20.3be836961cF8To&gatewayAdapt=glo2deu and this screen https://de.aliexpress.com/item/1005004152737300.html?spm=a2g0o.order_list.order_list_main.15.3be836961cF8To&gatewayAdapt=glo2deu 
Hope to receive asap so i can continue with my project....this time with TFT_espi

**Comment by Pconti31** [comment]
@yesnoj I just finished updating my XProgress control that supports text inside it and avoids the flicker problem.  It now also works with   Adafruit_GFX library along with TFT_eSPI library.  Now you can use either.
check it out at [progress-percent](https://github.com/Pconti31/GUIslice-Solutions/tree/main/apps/progress-percent)
Paul--

**Reply by yesnoj** [reply]
Thanks Paul!I will try asap!
